tensorflow/c/kernels/histogram_summary_op.cc

```diff
@@ -0,0 +1,157 @@
+#include "tensorflow/core/framework/types.h"
+#include "tensorflow/core/framework/selective_registration.h"
+#include "tensorflow/core/platform/tstring.h"
+#include "tensorflow/core/platform/protobuf.h"
+#include "tensorflow/core/platform/default/logging.h"
+#include "tensorflow/core/platform/macros.h"
+#include "tensorflow/core/lib/histogram/histogram.h"
+
+// Wrappers to delete resources once the resource is out of scope. 
+struct Tensor_Wrapper { 
+  TF_Tensor* t; 
+  Tensor_Wrapper() : t(nullptr) {}
+  ~Tensor_Wrapper() { 
+    TF_DeleteTensor(t);
+  }
+};
+
+struct Status_Wrapper { 
+  TF_Status* s; 
+  Status_Wrapper() { 
+    s = TF_NewStatus(); 
+  }
+  ~Status_Wrapper() { 
+    TF_DeleteStatus(s);
+  }
+};
+
+// dummy functions used for kernel registration 
+static void* HistogramSummaryOp_Create(TF_OpKernelConstruction* ctx) {
+  return nullptr;
+}
+
+static void HistogramSummaryOp_Delete(void* kernel) {
+  return;
+}
+
+template<typename T>
+static void HistogramSummaryOp_Compute(void* kernel, TF_OpKernelContext* ctx) {
+  Tensor_Wrapper tags_wrapper; 
+  Tensor_Wrapper values_wrapper; 
+  Status_Wrapper status_wrapper;
+  TF_GetInput(ctx, 0, &tags_wrapper.t, status_wrapper.s);
+  if (TF_GetCode(status_wrapper.s) != TF_OK) { 
+    TF_OpKernelContext_Failure(ctx, status_wrapper.s);
+    return; 
+  }
+  TF_GetInput(ctx, 1, &values_wrapper.t, status_wrapper.s);
+  if (TF_GetCode(status_wrapper.s) != TF_OK) { 
+    TF_OpKernelContext_Failure(ctx, status_wrapper.s);
+    return; 
+  }
+  if (TF_NumDims(tags_wrapper.t) != 0) { 
+    TF_SetStatus(status_wrapper.s, TF_INVALID_ARGUMENT, "tags must be scalar");
+    TF_OpKernelContext_Failure(ctx, status_wrapper.s);
+    return; 
+  }
+  // Cast values to array to access tensor elements by index 
+  auto values_array = static_cast<T*>(TF_TensorData(values_wrapper.t)); 
+  tensorflow::histogram::Histogram histo; 
+  for (int i = 0; i < TF_TensorElementCount(values_wrapper.t); ++i) { 
+    const double double_val = static_cast<double>(values_array[i]); 
+    if (Eigen::numext::isnan(double_val)) { 
+      std::ostringstream err; 
+      err << "Nan in summary histogram for: "; 
+      TF_SetStatus(status_wrapper.s, TF_INVALID_ARGUMENT, err.str().c_str());
+      break;
+    }
+    else if (Eigen::numext::isinf(double_val)) { 
+      std::ostringstream err; 
+      err << "Infinity in Histogram for: "; 
+      TF_SetStatus(status_wrapper.s, TF_INVALID_ARGUMENT, err.str().c_str());
+      break; 
+    }
+    histo.Add(double_val);
+  }
+  tensorflow::Summary s; 
+  tensorflow::Summary::Value* v = s.add_value(); 
+  const tensorflow::tstring& tag = *(static_cast<tensorflow::tstring*>(
+  		TF_TensorData(tags_wrapper.t))); 
+  v->set_tag(tag.data(), tag.size()); 
+  histo.EncodeToProto(v->mutable_histo(), false /* Drop zero buckets */); 
+
+  // Must use new status for TF_AllocateOutput if previous status is set 
+  // because of an invalid values argument.  
```

Comment:
I don't quite understand this comment. What do you mean by "invalid values argument"?

Reply:
In the C++ kernel, the program continues execution after an invalid_argument status, here. https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/kernels/summary_op.cc;l=94;drc=94e37f84f19384e685420ef7f90382fcfe719498

Because TF_AllocateOutput sets the status to TF_OK, https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/kernels.cc;l=254;drc=791f4f2b773208b21ece71e1d37204a91720346d, I use a new status so TF_AllocateOutput doesn't overwrite the TF_InvalidArugment Status

Reply:
hm wonder if that was originally intentional or a bug

Reply:
I'm not sure either. It seems more likely to be a bug though. Other kernels in core return after invalid_argument, and it doesn't quite make sense to output an incomplete histogram summary. How should I proceed?

Reply:
Let's return on failure. I checked a few places in the code and seems like outputs are not considered if context has a failure status. For example, here: https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/common_runtime/eager/kernel_and_device.cc;l=309?q=tensorflow%2Fcore%2Fcommon_runtime%2Feager%2Fkernel_and_device.cc

If this causes a test failure then we can reconsider.